### ISSUE TITLE
Add ports inventory for Mac OS X

### ISSUE THREAD

**Issue by Skeptor** [wazuh/wazuh#3296]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.11.0 | Syscollector | Agent | Packages/Sources | Mac OS X |

This issue is derived from #2923 

Add the ports inventory feature to the Mac OS agents' scan as it is done in Linux and Windows.
As first approach I've tested a simple port scan:
```
	//Initialise the sockaddr_in structure
	strncpy((char*)&sa , "" , sizeof sa);
	sa.sin_family = AF_INET;
	
	//direct ip address, use it
	if(isdigit(hostname[0]))
	{
		printf("Doing inet_addr...");
		sa.sin_addr.s_addr = inet_addr(hostname);
		printf("Done\n");
	}
	//Resolve hostname to ip address
	else if( (host = gethostbyname(hostname)) != 0)
	{
		printf("Doing gethostbyname...");
		strncpy((char*)&sa.sin_addr , (char*)host->h_addr , sizeof sa.sin_addr);
		printf("Done\n");
	}
	else
	{
		herror(hostname);
		exit(2);
	}
	
	//Start the port scan loop
	printf("Starting the portscan loop : \n");
	for( i = start ; i <= end ; i++) 
	{
		//Fill in the port number
		sa.sin_port = htons(i);
		//Create a socket of type internet
		sock = socket(AF_INET , SOCK_STREAM , 0);
		
		//Check whether socket created fine or not
		if(sock < 0) 
		{
			perror("\nSocket");
			exit(1);
		}
		//Connect using that socket and sockaddr structure
		err = connect(sock , (struct sockaddr*)&sa , sizeof sa);
		
		//not connected
		if( err < 0 )
		{
			printf("%s %-5d %s\r" , hostname , i, strerror(errno));
			fflush(stdout);
		}
		//connected
		else
		{
			printf("%-5d open\n",  i);
		}
		close(sock);
	}
	
	printf("\r");
	fflush(stdout);
	return(0);
```
It returns the port, the state (open or closed) and Connection refused in the case that the port is open and used by a process.

```
Starting the portscan loop : 
22    open74 21    Connection refused
514   open74 513   Connection refused
902   open74 901   Connection refused
1515  open74 1514  Connection refused
```

The issue with the scan is that it takes several  minutes on scan the MAC ports and it gives insufficient information on the open ports.
The task is to reduce the scan time and to extract all the information needed of each port.

Tasks
--------
- [x] Test another scan approach where the PIDs of the system are scanned to see if they are using any port.
   - [x] In case this scan doesn't work reduce the scan time and get the information from the socket connection
- [x] Test the solution in BSD systems to see if it's correct or it the code needs to be isolated to use it only on Mac OS.
- [x] Memory status. Run valgrind on `wazuh-modulesd`

**Comment by Skeptor** [comment]
I'm testing the second approach to scan the ports of the system. I'm using as reference some of the code of `lsof` https://opensource.apple.com/source/lsof/lsof-33/lsof/dialects/darwin/libproc/dsock.c.auto.html to obtain from the PIDs the information about the sockets they are using to know if they have a connection open and extract the port from there.

**Comment by Skeptor** [comment]
First approach to the port scan:
```
pid_t * pids = calloc(0x1000, 1);
    int count = proc_listallpids(pids, 0x1000);

    int index;

    for(index=0; index < count; ++index) {
        pid_t pid = pids[index];
        // Figure out the size of the buffer needed to hold the list of open FDs
        int bufferSize = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, 0, 0);
        if (bufferSize == -1) {
                printf("Unable to get open file handles for %d\n", pid);
                return 1;
        }

        // Get the list of open FDs
        struct proc_fdinfo *procFDInfo = (struct proc_fdinfo *)malloc(bufferSize);
        if (!procFDInfo) {
                printf("Out of memory. Unable to allocate buffer with %d bytes\n", bufferSize);
                return 1;
        }
        proc_pidinfo(pid, PROC_PIDLISTFDS, 0, procFDInfo, bufferSize);
        int numberOfProcFDs = bufferSize / PROC_PIDLISTFD_SIZE;

        int i;
	    int alertado = 0;
        for(i = 0; i < numberOfProcFDs; i++) {
            if(procFDInfo[i].proc_fdtype == PROX_FDTYPE_SOCKET) {
                // A socket is open
                struct socket_fdinfo socketInfo;
                int bytesUsed = proc_pidfdinfo(pid, procFDInfo[i].proc_fd, PROC_PIDFDSOCKETINFO, &socketInfo, PROC_PIDFDSOCKETINFO_SIZE);
                if (bytesUsed == PROC_PIDFDSOCKETINFO_SIZE) {
                    if(socketInfo.psi.soi_family == AF_INET && socketInfo.psi.soi_kind == SOCKINFO_TCP) {
			            int localPort = (int)ntohs(socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_lport);
                        int remotePort = (int)ntohs(socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_fport);
                        if (remotePort == 0 && localPort != 0) {
			                if(!alertado){
			    	            alertado = 1;
				                printf("\nOpen ports by PID: %d\n", pid);
			                }
                            // Remote port will be 0 when the FD represents a listening socket
                            printf("port IPv4: %d   ", localPort);
                        }
                    }
                } else if(socketInfo.psi.soi_family == AF_INET6 && socketInfo.psi.soi_kind == SOCKINFO_TCP){
                    int localPort = (int)ntohs(socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_lport);
                    int remotePort = (int)ntohs(socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_fport);
                    if (remotePort == 0) {
                                    if(!alertado){
                                        alertado = 1;
                                        printf("\nOpen ports by PID: %d\n", pid);
                                    }
                                    // Remote port will be 0 when the FD represents a listening socket
                                    printf("port IPv6: %d   ", localPort);
                    }
                }
            }
        }
    }
```

This scan returns the ports opened by a process with the protocol (IPv4 or IPv6)

```
Open ports by PID: 50988
port IPv4: 5930    port IPv4: 2200    
Open ports by PID: 50186
port IPv4: 5944    port IPv4: 2222    
Open ports by PID: 98332
port IPv4: 2202    
Open ports by PID: 55604
port IPv4: 63981    port IPv6: 63981     
Open ports by PID: 30511
port IPv6: 80     
Open ports by PID: 29625
port IPv4: 59079    
Open ports by PID: 19906
port IPv4: 2201    
Open ports by PID: 12570
port IPv6: 631     port IPv4: 631    
Open ports by PID: 398
port IPv4: 49166    port IPv6: 49166     
Open ports by PID: 236
port IPv6: 80     
Open ports by PID: 100
port IPv6: 88     
Open ports by PID: 88
port IPv6: 80     
Open ports by PID: 65
port IPv4: 5951    
Open ports by PID: 1
port IPv6: 22     port IPv6: 22     port IPv4: 22    port IPv6: 5900     port IPv4: 5900    port IPv4: 22    port IPv6: 5900     port IPv4: 5900    
```

The next step is retrieve the local and remote IPs, the queue values, the inode and the state

**Comment by Skeptor** [comment]
The scan now obtain the state of the port too from the variable `tcpsi_state`
```
Open ports by PID: 56480
port IPv4: 5944     listening 
port IPv4: 2222     listening 

Open ports by PID: 50988
port IPv4: 5930     listening 
port IPv4: 2200     listening 

Open ports by PID: 98332
port IPv4: 2202     listening 

Open ports by PID: 55604
port IPv4: 63981     listening 
port IPv6: 63981      listening 

Open ports by PID: 30511
port IPv6: 80      listening 

Open ports by PID: 29625
port IPv4: 59079     listening 

Open ports by PID: 19906
port IPv4: 2201     listening 

Open ports by PID: 12570
port IPv6: 631      listening 
port IPv4: 631     listening 

Open ports by PID: 398
port IPv4: 49166     listening 
port IPv6: 49166      listening 

Open ports by PID: 236
port IPv6: 80      listening 

Open ports by PID: 100
port IPv6: 88      listening 

Open ports by PID: 88
port IPv6: 80      listening 

Open ports by PID: 65
port IPv4: 5951     listening 

Open ports by PID: 1
port IPv6: 22      listening 
port IPv6: 22      listening 
port IPv4: 22     listening 
port IPv6: 5900      listening 
port IPv4: 5900     listening 
port IPv4: 22     listening 
port IPv6: 5900      listening 
port IPv4: 5900     listening 
```

I'm trying to retrieve the local addres and the remote addres from `socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_laddr.ina_46.i46a_addr4` and `socketInfo.psi.soi_proto.pri_tcp.tcpsi_ini.insi_faddr.ina_46.i46a_addr4`.
Also i'm investigating how to know the queue values for each connection

**Comment by Skeptor** [comment]
Update of the scan with UDP, local address and remote address
```
Open ports by PID: 70
port TCP: 5951    Local Address 127.0.0.1  Remote Address 0.0.0.0  listening

Open ports by PID: 68
port UDP6: 0     Local address :: Remote address ::

Open ports by PID: 55
port UDP: 28127    Local Address 0.0.0.0  Remote Address 0.0.0.0  

Open ports by PID: 1
port TCP6: 5900     Local address :: Remote address ::  listening
port TCP: 5900    Local Address 0.0.0.0  Remote Address 0.0.0.0  listening
port TCP6: 22     Local address :: Remote address ::  listening
port TCP: 22    Local Address 0.0.0.0  Remote Address 0.0.0.0  listening
port TCP6: 22     Local address :: Remote address ::  listening
port TCP: 22    Local Address 0.0.0.0  Remote Address 0.0.0.0  listening
port UDP: 35328    Local Address 0.0.0.0  Remote Address 0.0.0.0  
port TCP6: 5900     Local address :: Remote address ::  listening
port TCP: 5900    Local Address 0.0.0.0  Remote Address 0.0.0.0  listening
port UDP: 35072    Local Address 0.0.0.0  Remote Address 0.0.0.0  
```

**Comment by Skeptor** [comment]
I'm working in the branch https://github.com/wazuh/wazuh/tree/issue-3296

The last commit https://github.com/wazuh/wazuh/commit/efea22e11592facf5c4651278f3f3028e1039927 makes the port scan of syscollector available for MAC agents.

This is the information stored in the DB with the option `scan_all` activated
```
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |49194|192.168.0.74  |1514||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |35072|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |35328|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|tcp|127.0.0.1  |49152|127.0.0.1  |5951||||established||
698580644|2019/05/17 17:40:21|tcp6|::|80|::|0||||listening||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |59668|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|udp6|::|5353|::|0||||||
698580644|2019/05/17 17:40:21|udp6|::|88|::|0||||||
698580644|2019/05/17 17:40:21|tcp6|::|88|::|0||||listening||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |22528|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|tcp|0.0.0.0  |88|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|tcp6|::|80|::|0||||listening||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |49153|17.252.76.99  |5223||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |49153|17.252.76.99  |5223||||established||
698580644|2019/05/17 17:40:21|tcp|127.0.0.1  |5951|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|tcp|127.0.0.1  |5951|127.0.0.1  |49152||||established||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |49158|37.252.246.105  |5938||||established||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |28127|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|tcp6|::|5900|::|0||||listening||
698580644|2019/05/17 17:40:21|tcp|0.0.0.0  |5900|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|tcp6|::|22|::|0||||listening||
698580644|2019/05/17 17:40:21|tcp|0.0.0.0  |22|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|tcp6|::|22|::|0||||listening||
698580644|2019/05/17 17:40:21|tcp|0.0.0.0  |22|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |35328|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|tcp6|::|5900|::|0||||listening||
698580644|2019/05/17 17:40:21|tcp|0.0.0.0  |5900|0.0.0.0  |0||||listening||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
698580644|2019/05/17 17:40:21|udp|0.0.0.0  |35072|0.0.0.0  |0||||||
698580644|2019/05/17 17:40:21|tcp|192.168.0.253  |22|192.168.0.74  |59728||||established||
```
And in Kibana it looks like this:
![image](https://user-images.githubusercontent.com/9034923/57939944-b745a600-78cb-11e9-95ca-a5c9cceabe52.png)

![image](https://user-images.githubusercontent.com/9034923/57939955-bf9de100-78cb-11e9-9098-ee49be81e0b3.png)

The next step is change the `ossec.conf` template of `darwin` to set the port scan enabled in the default configuration and run test to ensure the funcionality is solid.

**Comment by chemamartinez** [comment]
Merged #3349
